File: Android/TUIKit/TUICallKit/tuicallkit/src/main/java/com/tencent/qcloud/tuikit/tuicallkit/extensions/CallingKeepAliveFeature.java
package com.tencent.qcloud.tuikit.tuicallkit.extensions;

import android.content.Context;

import com.tencent.qcloud.tuikit.tuicallkit.service.TUICallService;
import com.tencent.qcloud.tuikit.tuicallkit.utils.DeviceUtils;

public class CallingKeepAliveFeature {
    private Context mContext;
    private boolean mEnableKeepAlive = true;

    public CallingKeepAliveFeature(Context context) {
        mContext = context;
    }

    public void enableKeepAlive(boolean enable) {
        mEnableKeepAlive = enable;
    }

    public void startKeepAlive() {
        if (!mEnableKeepAlive) {
            return;
        }
        TUICallService.start(mContext);
    }

    public void stopKeepAlive() {
        if (DeviceUtils.isServiceRunning(mContext, TUICallService.class.getName())) {
            TUICallService.stop(mContext);
        }
    }
}

File: Android/TUIKit/TUICallKit/tuicallkit-kt/src/main/java/com/tencent/qcloud/tuikit/tuicallkit/extensions/CallingKeepAliveFeature.kt
package com.tencent.qcloud.tuikit.tuicallkit.extensions

import android.content.Context
import com.tencent.qcloud.tuikit.tuicallengine.TUICallDefine
import com.tencent.qcloud.tuikit.tuicallengine.impl.base.Observer
import com.tencent.qcloud.tuikit.tuicallkit.state.TUICallState
import com.tencent.qcloud.tuikit.tuicallkit.utils.DeviceUtils

class CallingKeepAliveFeature(private val mContext: Context) {
    init {
        addObserver()
    }

    private var callStatusObserver = Observer<TUICallDefine.Status> {
        if (it == TUICallDefine.Status.None) {
            stopKeepAlive()
        } else if (it == TUICallDefine.Status.Waiting) {
            startKeepAlive()
        }
    }

    fun addObserver() {
        TUICallState.instance.selfUser.get().callStatus.observe(callStatusObserver)
    }

    private fun startKeepAlive() {
        TUICallService.start(mContext)
    }

    private fun stopKeepAlive() {
        TUICallState.instance.selfUser.get().callStatus.removeObserver(callStatusObserver)
        if (DeviceUtils.isServiceRunning(mContext, TUICallService::class.java.name)) {
            TUICallService.stop(mContext)
        }
    }
}